styx-scheduler-service/src/main/java/com/spotify/styx/docker/KubernetesDockerRunner.java

```diff
@@ -504,124 +507,51 @@ public void close() throws IOException {
-        runState.state().equals(RUNNING) && runState.data().executionId().isPresent());
-
-    final Set<WorkflowInstance> workflowInstancesForPods = podList.getItems().stream()
-        .map(pod -> pod.getMetadata().getAnnotations())
-        .filter(Objects::nonNull)
-        .map(annotations -> annotations.get(STYX_WORKFLOW_INSTANCE_ANNOTATION))
-        .filter(Objects::nonNull)
-        .map(WorkflowInstance::parseKey)
-        .collect(toSet());
-
-    // Emit errors for workflow instances that seem to be missing its pod
-    runningWorkflowInstances.forEach((workflowInstance, runState) -> {
-
-      // Is there a matching pod in the list? Bail.
-      if (workflowInstancesForPods.contains(workflowInstance)) {
-        return;
-      }
-
-      // The pod list might be stale so explicitly look for a pod using the execution ID.
-      final String executionId = runState.data().executionId().get();
-      final Pod pod = client.pods().withName(executionId).get();
-
-      // We found a pod? Bail.
-      if (pod != null) {
-        return;
-      }
-
-      // No pod found. Emit an error guarded by the state counter we are basing the error conclusion on.
-      stateManager.receiveIgnoreClosed(
-          Event.runError(workflowInstance, "No pod associated with this instance"), runState.counter());
-    });
-  }
-
-  private void pollPods() {
+  private void cleanupPods() {
     try {
-      try (Scope ss = tracer.spanBuilder("Styx.KubernetesDockerRunner.pollPods")
+      try (Scope ss = tracer.spanBuilder("Styx.KubernetesDockerRunner.cleanupPods")
           .setRecordEvents(true)
           .setSampler(Samplers.alwaysSample())
           .startScopedSpan()) {
-        tryPollPods();
+        tryCleanupPods();
       }
     } catch (Throwable t) {
-      LOG.warn("Error while polling pods", t);
+      LOG.warn("Error while cleaning pods", t);
     }
   }
 
+  /**
+   * Deletes stale workflow instance execution pods.
+   */
   @VisibleForTesting
-  synchronized void tryPollPods() {
-    // Fetch pods _before_ fetching all active states
-    final PodList list = client.pods().list();
-    final Map<WorkflowInstance, RunState> activeStates = stateManager.getActiveStates();
-      
-    // Note: Avoid blocking during the processing of the above fetched run states in order to prevent them
-    //       from going stale and losing races due to concurrent processing with other schedulers.
-      
-    examineRunningWFISandAssociatedPods(activeStates, list);
-
-    var podsToDelete = new ArrayList<String>();
-    for (Pod pod : list.getItems()) {
-      logEvent(Watcher.Action.MODIFIED, pod, list.getMetadata().getResourceVersion(), true);
-      final Optional<WorkflowInstance> workflowInstance = readPodWorkflowInstance(pod);
-      if (!workflowInstance.isPresent()) {
-        continue;
-      }
+  void tryCleanupPods() {
+    client.pods().list().getItems().stream()
+        .map(pod -> runAsync(guard(() -> tryCleanupPod(pod)), eventExecutor))
```

Comment:
Shall we have a new dedicated executor so cleanup will not compete with even handling?

Reply:
I think it's fine to share the executor for cleanup and pod update processing. It won't block the watcher.